### DISCUSSION TITLE
How can I automatically sign-in the user on server? (without requiring them to do anything like clicking an email link)

### DISCUSSION THREAD

**Discussion by lumenwrites** [supabase/supabase#19320] [Questions] [answered]
Hi! I'm implementing payments for my app using Next.js, Supabase, and Stripe. My goal is to implement the following purchase flow:
- New user visits the website, enters their email and credit card info to make a purchase.
- After the purchase, stripe redirects them to the "Thank you!" page, where I use the email they've entered to create them a new account, and sign them in (I want it all happen automatically, on the server side, before the page loads).
- When the user sees the loaded "Thank you!" page, they're already signed in, and are looking at a page where they can download the purchased product.

I can create a user account like this:

```
export async function handlePurchase({ email }) {
  const cookieStore = cookies()
  const supabase = createServerClient(cookieStore)
  const { data, error } = await supabase.auth.admin.createUser({
    email,
    email_confirm: true,
  })
  console.log('[handlePurchase] createUser', { data, error })
  // Log in the created user
  ???
}
```

But how can I automatically log them in after the account has been created?

**Comment by GaryAustin1** [comment, marked as answer]
If no one comes along with a better way... You might use admin call generateEmailLink for magic link.   Then it will give you tokens you would normally send to the user.  You can use at least the 6 digit token with verifyOtp to signin the user.   I think, but have never tried it.  Once signed in you would have their access token and refresh token.

**Reply by lumenwrites** [reply]
Thank you so much for your help! This seems to work!

For anyone googling this later, here's my code (for my Next.js app).

On server:
```
async function signUpOrSignIn(email) {
  const supabase = createServerClient()
  // Generate a magic link, email_otp, and hashed_token so I could use one of them to automatically log user in
  // https://supabase.com/docs/reference/javascript/auth-admin-generatelink
  const { data: generateLinkData, error: generateLinkError } =
    await supabase.auth.admin.generateLink({
      email,
      type: 'magiclink',
    })
  return {
    generateLinkTokens: generateLinkData.properties,
    user: generateLinkData.user,
  }
}
```

On client:
```
export default function SignInOnClient({ hashed_token }) {
  useEffect(() => {
    async function signIn() {
      const supabase = createBrowserClient()
      const { data: verifyOtpData, error: verifyOtpError } =
        await supabase.auth.verifyOtp({
          token_hash: hashed_token,
          type: 'email',
        })
      console.log('[SignInOnClient]', { verifyOtpData, verifyOtpError })
      const { data, error } = await supabase.auth.getSession()
      console.log('[SignInOnClient] Supabase session:', { data, error })
    }
    signIn()
  }, [])
  return <div>Yes</div>
}
```

(hashed token is under `generateLinkData.properties.hashed_token`)

**Reply by wiesson** [reply]
What happens if the email address already exists? I have the same use-case as you have, but I already have like 2k users. I thought about searching the auth.users by stripes customer email address and do the login on the server, but then I was thinking that any user (with money) can take over other accounts. 

I assumed that if somebody uses their creditcard and pays like 100 USD/y, (min plan in my app) it should be same as confirming the email address.

**Reply by wiesson** [reply]
After thinking about that, if the email exists, I redirect the user to another page and send out the email login mail to make sure, the user has access to it.

**Reply by wiesson** [reply]
Btw, with SSR, it's also not complicated:

```Node
export async function POST(req: Request) {
  const url = new URL(req.url);
  const originUrl = `${url.protocol}//${url.host}`;

  const { data: generateLinkData } = await adminClient.auth.admin.generateLink({
    email: userEmail,
    type: "magiclink"
  });

  const supabase = createUserServerClient();
  const { error } = await supabase.auth.verifyOtp({
    type: "magiclink",
    token_hash: generateLinkData.properties.hashed_token,
  });

  return NextResponse.redirect(`${originUrl}/membership/welcome`);
}
```

and done :)
